### DISCUSSION TITLE
Generate tokens result for "parses_to" macro?

### DISCUSSION THREAD

**Discussion by TiemenSch** [pest-parser/pest#877] [Q&A]
While the `parses_to! {}` macro makes debugging parser output very nice as soon as you get the tokens setup, I would very much like to easily update the expected output with what I'm actually getting.

Currently it errors with by enumerating a whole lot of `Start` and `End` "events", but I can't just copy-paste that into the tokens list after I re-adjust my grammar to capture the right things.

Even for quite little examples, it will get wordy and indexing is a hassle to update:

Do you know of any way to generate the nested tokens list and print it?

```rust
    #[test]
    fn parse_relation_definition() {
        parses_to! {
            parser: FooParser,
            input:
r#"define relation foo
relating parameters
one or more bar are a baz
"#,
            rule: Rule::relation_definition,
            // How to update this list easily?
            tokens: [relation_definition(0, 65, [
                relation_definition_header(0, 19, [
                    kw_define(0, 6),
                    kw_relation(7, 15),
                    some_definition(16, 19, [
                        name(16, 19)
                    ]),
                ]),
                relating_section(20, 65, [
                    kw_relating(20, 28),
                    kw_parameters(29, 39),
                    relation_parameter_line(40, 65, [
                        kw_one_or_more(40, 51),
                        some_variable(52, 55, [
                            name(52, 55)
                        ]),
                        kw_is_are(56, 59),
                        kw_a(60, 61),
                        some_type(62, 65, [
                            name(62, 65)
                        ])
                    ])
                ])
            ])]
        };
    }
```

**Comment by steadygaze** [comment]
After some messing around, I found that the [regular `Display.fmt` implementation for Pairs](https://docs.rs/pest/2.7.10/pest/iterators/struct.Pairs.html#method.fmt) corresponds closely to the correct output for `tokens`. Therefore, you can simply run

```rust
println!("{}", FooParser::parse(Rule::relation_definition, "your input").unwrap());
```

and copy/paste the output into `tokens`. The main issue is that it's all on one line and not formatted prettily, so you have to do that yourself. It is also annoying to have to print out the value manually in the first place. So I'd say there is room for improvement in pest in this area.
